simpledb/query/sort_scan.go

```diff
@@ -0,0 +1,166 @@
+package query
+
+import (
+	"fmt"
+	"simpledb/record"
+)
+
+var _ Scan = (*SortScan)(nil)
+
+type SortScan struct {
+	s1, s2             UpdateScan
+	currentScan        UpdateScan
+	comp               *RecordComparator
+	hasMore1, hasMore2 bool
+	savedPosition      []*record.RID
+}
+
+func NewSortScan(runs []*TempTable, comp *RecordComparator) (*SortScan, error) {
+	s1, err := runs[0].Open()
+	if err != nil {
+		return nil, fmt.Errorf("runs[0].Open: %w", err)
+	}
+	hasMore1, err := s1.Next()
+	if err != nil {
+		return nil, fmt.Errorf("s1.Next: %w", err)
+	}
+
+	var s2 UpdateScan
+	var hasMore2 bool
+	if len(runs) > 1 {
+		s2, err = runs[1].Open()
+		if err != nil {
+			return nil, fmt.Errorf("runs[1].Open: %w", err)
+		}
+		hasMore2, err = s2.Next()
+		if err != nil {
+			return nil, fmt.Errorf("s2.Next: %w", err)
+		}
+	}
+
+	return &SortScan{
+		s1:          s1,
+		s2:          s2,
+		currentScan: nil,
+		comp:        comp,
+		hasMore1:    hasMore1,
+		hasMore2:    hasMore2,
+	}, nil
+}
+
+func (ss *SortScan) BeforeFirst() error {
+	var err error
+	if err = ss.s1.BeforeFirst(); err != nil {
+		return fmt.Errorf("ss.s1.BeforeFirst: %w", err)
+	}
+	ss.hasMore1, err = ss.s1.Next()
+	if err != nil {
+		return fmt.Errorf("ss.s1.Next: %w", err)
+	}
+
+	if ss.s2 != nil {
+		if err = ss.s2.BeforeFirst(); err != nil {
+			return fmt.Errorf("ss.s2.BeforeFirst: %w", err)
+		}
+		ss.hasMore2, err = ss.s2.Next()
+		if err != nil {
+			return fmt.Errorf("ss.s2.Next: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (ss *SortScan) Next() (bool, error) {
+	var err error
+
+	if ss.currentScan == ss.s1 {
+		ss.hasMore1, err = ss.s1.Next()
+		if err != nil {
+			return false, fmt.Errorf("ss.s1.Next: %w", err)
+		}
+	} else if ss.s2 != nil && ss.currentScan == ss.s2 {
+		ss.hasMore2, err = ss.s2.Next()
+		if err != nil {
+			return false, fmt.Errorf("ss.s2.Next: %w", err)
+		}
+	}
+
+	if !ss.hasMore1 && !ss.hasMore2 {
+		return false, nil
+	} else if ss.hasMore1 && ss.hasMore2 {
+		cmp, err := ss.comp.Compare(ss.s1, ss.s2)
+		if err != nil {
+			return false, fmt.Errorf("ss.comp.Compare: %w", err)
+		}
+		if cmp < 0 {
+			ss.currentScan = ss.s1
+		} else {
+			ss.currentScan = ss.s2
+		}
+	} else if ss.hasMore1 {
+		ss.currentScan = ss.s1
+	} else if ss.hasMore2 {
+		ss.currentScan = ss.s2
+	}
+
+	return true, nil
```

Comment:
[nits][imo]

ネスト減らせないかなぁとおもいましたが、否定のif文だらけなのどっちが認知不可低いか微妙なところ…

```suggestion
	if !ss.hasMore1 && !ss.hasMore2 {
		return false, nil
	}

	if !ss.hasMore2  {
		ss.currentScan = ss.s1
		return true, nil
	}

	if !ss.hasMore1  {
		ss.currentScan = ss.s2
		return true, nil
	}
	
	cmp, err := ss.comp.Compare(ss.s1, ss.s2)
	if err != nil {
		return false, fmt.Errorf("ss.comp.Compare: %w", err)
	}

	if cmp < 0 {
		ss.currentScan = ss.s1
	} else {
		ss.currentScan = ss.s2
	}

	return true, nil
```